apps/functions-subscription/src/use-cases/__tests__/insert-subscription.test.ts

```diff
@@ -0,0 +1,82 @@
+import { describe, expect, it } from 'vitest';
+import * as TE from 'fp-ts/TaskEither';
+import * as O from 'fp-ts/Option';
+import * as E from 'fp-ts/Either';
+import { insertSubscription } from '../insert-subscription';
+import {
+  aSubscription,
+  aSubscriptionRequest,
+} from '../../domain/__tests__/data';
+import { makeTestEnv } from '../../domain/__tests__/mocks';
+import { SubscriptionStoreError } from '../errors';
+import { ItemAlreadyExists } from '../../domain/errors';
+
+const { userId, trialId } = aSubscription;
+
+describe('insertSubscription', () => {
+  it('should return error if the subscription already exists', async () => {
+    const testEnv = makeTestEnv();
+
+    testEnv.subscriptionReader.get.mockReturnValueOnce(
+      TE.right(O.some(aSubscription)),
+    );
+    testEnv.hashFn.mockReturnValueOnce({ value: aSubscription.id });
+
+    const actual = await insertSubscription(userId, trialId)(testEnv)();
+    const expected = E.left(
+      new ItemAlreadyExists('Subscription already exists'),
+    );
+    expect(actual).toStrictEqual(expected);
+  });
+
+  it('should return the subscription created', async () => {
+    const testEnv = makeTestEnv();
+
+    testEnv.clock.now.mockReturnValue(aSubscription.createdAt);
+    testEnv.subscriptionReader.get.mockReturnValueOnce(TE.right(O.none));
+    testEnv.hashFn.mockReturnValueOnce({ value: aSubscription.id });
+    testEnv.subscriptionWriter.insert.mockImplementationOnce((_) =>
+      TE.right(_),
+    );
```

Comment:
**nitpick**: I would use `mockReturnValueOnce`.

As we already do, to enforce coherence I'd use that method to mock.

```suggestion
    testEnv.subscriptionWriter.insert.mockReturnValueOnce(
      TE.right(aSubscription),
    );
```